File: app/models/adjustment_bridge.py
"""Adjustment Bridge — pivot/data-cube workpaper model (Tier 1.9 skeleton).

This model stores pre-computed rows for the Adjustment Bridge cube.
Each row represents one account in one period/scenario combination,
with imported vs. adjusted balances and all pivot dimensions.

The full cube is computed on-demand from journal entries + imported balances.
Saved views, slicers, and groupings are stored in AdjustmentBridgeView.
"""
from sqlalchemy import Column, DateTime, ForeignKey, Integer, Numeric, String, Text, Boolean
from sqlalchemy.sql import func
from app.database import Base


class AdjustmentBridgeRow(Base):
    """One row in the Adjustment Bridge cube.

    Dimensions (grouping axes):
      entity_id, period_end, scenario_id, account_id
      account_type, taxonomy_category, fs_line
      adjustment_type, is_posted, is_included
      source, consolidation_group

    Measures (numeric values):
      imported_balance, posted_adjustments, draft_adjustments
      excluded_adjustments, adjusted_balance, variance
      prior_period_balance (nullable — requires prior period import)
      budget_placeholder (nullable — future)
    """

    __tablename__ = "adjustment_bridge_rows"

    id = Column(Integer, primary_key=True)

    # Core dimensions
    entity_id = Column(Integer, ForeignKey("entities.id"), nullable=False)
    period_end = Column(String(20), nullable=False)           # YYYY-MM-DD
    scenario_id = Column(Integer, ForeignKey("scenarios.id"), nullable=True)
    account_id = Column(Integer, ForeignKey("accounts.id"), nullable=True)
    account_number = Column(String(50), nullable=True)
    account_name = Column(String(200), nullable=True)
    account_type = Column(String(30), nullable=True)          # asset|liability|equity|revenue|expense
    taxonomy_category = Column(String(50), nullable=True)     # reporting taxonomy code
    fs_line = Column(String(100), nullable=True)              # financial statement line label

    # Pivot dimensions
    adjustment_type = Column(String(50), nullable=True)       # audit_adj|topside|elimination|accrual|pro_forma|tax
    is_posted = Column(Boolean, nullable=False, default=False)
    is_included = Column(Boolean, nullable=False, default=True)
    source = Column(String(50), nullable=True)                # pdf_import|coa_import|manual|system
    consolidation_group = Column(String(50), nullable=True)

    # Measures
    imported_balance = Column(Numeric(18, 2), nullable=True)
    posted_adjustments = Column(Numeric(18, 2), nullable=True)
    draft_adjustments = Column(Numeric(18, 2), nullable=True)
    excluded_adjustments = Column(Numeric(18, 2), nullable=True)
    adjusted_balance = Column(Numeric(18, 2), nullable=True)
    variance = Column(Numeric(18, 2), nullable=True)
    prior_period_balance = Column(Numeric(18, 2), nullable=True)
    budget_placeholder = Column(Numeric(18, 2), nullable=True)

    # Audit
    computed_at = Column(DateTime, nullable=False, server_default=func.now())
    notes = Column(Text, nullable=True)


class AdjustmentBridgeView(Base):
    """Saved view configuration for the Adjustment Bridge cube.

    Stores user-defined pivot layout: which dimensions are rows/columns/slicers,
    active filters, and sort order. One entity can have multiple named views.
    """

    __tablename__ = "adjustment_bridge_views"

    id = Column(Integer, primary_key=True)
    entity_id = Column(Integer, ForeignKey("entities.id"), nullable=False)
    name = Column(String(100), nullable=False)
    description = Column(Text, nullable=True)
    is_default = Column(Boolean, nullable=False, default=False)

    # JSON-encoded pivot config
    row_dimensions = Column(Text, nullable=True)     # ["account_type", "taxonomy_category"]
    column_dimensions = Column(Text, nullable=True)  # ["scenario_id"]
    slicer_config = Column(Text, nullable=True)      # {"entity_id": 1, "period_end": "2024-12-31"}
    sort_config = Column(Text, nullable=True)        # [{"field": "adjusted_balance", "dir": "desc"}]
    visible_measures = Column(Text, nullable=True)   # ["imported_balance", "adjusted_balance", "variance"]

    created_at = Column(DateTime, nullable=False, server_default=func.now())
    updated_at = Column(DateTime, nullable=False, server_default=func.now())

File: frontend/src/api/adjustmentBridge.ts
import api from './client'

export interface JeImpactResult {
  ni_impact: number
  bs_impact: number
  asset_impact: number
  liability_impact: number
  equity_impact: number
  line_details: Array<{
    account_id: number
    account_number: string
    account_name: string
    account_type: string
    debit: number
    credit: number
    accounting_balance: number
    presentation_amount: number
  }>
}

export interface AdjustmentBridgeRow {
  id: number
  entity_id: number
  period_end: string
  scenario_id: number | null
  account_id: number | null
  account_number: string | null
  account_name: string | null
  account_type: string | null
  taxonomy_category: string | null
  fs_line: string | null
  adjustment_type: string | null
  is_posted: boolean
  is_included: boolean
  source: string | null
  consolidation_group: string | null
  imported_balance: string | null
  posted_adjustments: string | null
  draft_adjustments: string | null
  excluded_adjustments: string | null
  adjusted_balance: string | null
  variance: string | null
  prior_period_balance: string | null
  budget_placeholder: string | null
  notes: string | null
}

export interface AdjustmentBridgeView {
  id: number
  entity_id: number
  name: string
  description: string | null
  is_default: boolean
  row_dimensions: string[] | null
  column_dimensions: string[] | null
  slicer_config: Record<string, unknown> | null
  sort_config: Array<{ field: string; dir: string }> | null
  visible_measures: string[] | null
}

export interface ComputeResult {
  entity_id: number
  period_end: string
  scenario_id: number | null
  rows_computed: number
  message: string
}

export interface CPABridgeColumn {
  je_id: number
  je_number: string
  description: string
  entry_date: string
}

export interface CPABridgeRow {
  account_id: number
  account_number: string
  account_name: string
  account_type: string
  account_sort: number
  as_reported: number
  ajes: Record<string, number>
  total_ajes: number
  adjusted: number
}

export interface CPABridgeResult {
  entity_id: number
  period_end: string
  scenario_id: number | null
  columns: CPABridgeColumn[]
  rows: CPABridgeRow[]
  totals: {
    as_reported: number
    ajes: Record<string, number>
    total_ajes: number
    adjusted: number
  }
}

export interface BridgeAdjustment {
  id: number
  je_number: string
  sequence: number
  entry_date: string
  description: string
  status: string
  total_debit: number
  total_credit: number
}

export interface BridgeRow {
  row_type: 'section' | 'account'
  level: number
  label: string
  account_id: number | null
  account_number: string | null
  account_name: string | null
  account_type: string | null
  as_reported: number
  adjustment_impacts: Record<string, number>
  total_ajes: number
  adjusted_balance: number
}

export interface BridgeResult {
  entity_id: number
  period_end: string
  scenario_id: number | null
  reporting_basis: string
  adjustments: BridgeAdjustment[]
  rows: BridgeRow[]
  totals: {
    as_reported: number
    adjustment_impacts: Record<string, number>
    total_ajes: number
    adjusted_balance: number
  }
}

export const adjustmentBridgeApi = {
  compute: (entityId: number, periodEnd: string, scenarioId?: number): Promise<ComputeResult> =>
    api.get<ComputeResult>('/adjustment-bridge/compute', {
      params: { entity_id: entityId, period_end: periodEnd, scenario_id: scenarioId },
    }).then((r) => r.data),

  rows: (params: {
    entity_id: number
    period_end?: string
    scenario_id?: number
    account_type?: string
    taxonomy_category?: string
  }): Promise<AdjustmentBridgeRow[]> =>
    api.get<AdjustmentBridgeRow[]>('/adjustment-bridge/rows', { params }).then((r) => r.data),

  listViews: (entityId: number): Promise<AdjustmentBridgeView[]> =>
    api.get<AdjustmentBridgeView[]>('/adjustment-bridge/views', { params: { entity_id: entityId } }).then((r) => r.data),

  createView: (body: Partial<AdjustmentBridgeView> & { entity_id: number; name: string }): Promise<AdjustmentBridgeView> =>
    api.post<AdjustmentBridgeView>('/adjustment-bridge/views', body).then((r) => r.data),

  updateView: (viewId: number, body: Partial<AdjustmentBridgeView>): Promise<AdjustmentBridgeView> =>
    api.patch<AdjustmentBridgeView>(`/adjustment-bridge/views/${viewId}`, body).then((r) => r.data),

  deleteView: (viewId: number): Promise<void> =>
    api.delete(`/adjustment-bridge/views/${viewId}`).then(() => undefined),

  cpaBridge: (params: {
    entity_id: number
    period_end: string
    scenario_id?: number
  }): Promise<CPABridgeResult> =>
    api.get<CPABridgeResult>('/adjustment-bridge/cpa-bridge', { params }).then((r) => r.data),

  bridge: (params: {
    entity_id: number
    period_end: string
    scenario_id?: number
    reporting_basis?: string
  }): Promise<BridgeResult> =>
    api.get<BridgeResult>('/adjustment-bridge/bridge', { params }).then((r) => r.data),

  exportBridgeCsv: (entityId: number, periodEnd: string, scenarioId?: number, reportingBasis = 'adjusted'): Promise<Blob> =>
    api.get('/adjustment-bridge/bridge/export-csv', {
      params: { entity_id: entityId, period_end: periodEnd, scenario_id: scenarioId, reporting_basis: reportingBasis },
      responseType: 'blob',
    }).then((r) => r.data),

  getJeImpact: (jeId: number): Promise<JeImpactResult> =>
    api.get<JeImpactResult>(`/adjustment-workspace/journal-entries/${jeId}/impact`).then((r) => r.data),
}
